Rebase October CMS on Laravel 6

bennothommo opened the pull request:
Fixes #4381. Related Rain library PR: https://github.com/octobercms/library/pull/456. Related docs PR: https://github.com/octobercms/docs/pull/446

This PR represents the current progress for rebasing October CMS on the Laravel 6 framework. This PR can be considered an "early beta" release. It is not recommended for use on production installations, however, we would encourage users to test this out on development copies in preparation for the upgrade.

**[Release notes and required changes can be found here.](https://github.com/octoberrain/meta/blob/master/l6-upgrade-notes.md)**

## New minimum requirements
- PHP 7.2 or above.

## Testing
We would appreciate any assistance with testing from the community. If you are interested in testing this out, please follow the below steps to quickly get a fresh instance of the CMS running.

### Testing on existing projects

Follow the Upgrade Notes here: https://github.com/octoberrain/meta/blob/master/l6-upgrade-notes.md

### Testing on fresh instance

1. Download and unzip the Laravel 6 branch into a directory, or clone the repository and checkout the **wip/laravel-6** branch.
    - Download: https://github.com/octobercms/october/archive/wip/laravel-6.zip
    - Clone: `git clone git@github.com:octobercms/october.git`
2. Install the Composer dependencies within the newly created folder.
    - Run `composer install`
3. Generate a **.env** file.
    - Run `php artisan october:env`
4. Change the following details in your **.env** file.
    - **APP_URL**: `http://127.0.0.1:8000`
    - **APP_KEY**: Run `php artisan key:generate --show` and copy and paste the value generated here.
    - **DB_CONNECTION**: `sqlite`
    - **DB_DATABASE**: `storage/database.sqlite`
5. Create a new SQLite file in **storage/database.sqlite**.
    - Run `touch storage/database.sqlite`
6. Run all the migrations to set up the database.
    - Run `php artisan october:up`
7. Finally, serve the website.
    - Run `php artisan serve`

This will allow you to review the CMS in the following contexts:
- **Frontend:** http://127.0.0.1/
- **Backend:** http://127.0.0.1/backend (log in using "admin" as username, "admin1234" as password)

## Reporting bugs and issues
At this stage, do not create a new issue on GitHub if you spot any issues. Please report the issue as a comment to https://github.com/octobercms/october/issues/4381, which we are using to track progress on the task.

Commit:
Added storage/framework/cache/data folder required by 5.6.

Commit:
Config changes from 5.6

Commit:
Default backend user password requirements changed for utf8mb4 DB type and minimum password length requirements

Commit:
Changed calls to the Cache to use DateTime instances instead of integers representing minutes as 5.8 changed integers into meaning seconds instead.

Commit:
Replaced JSMin with JSqueeze

Commit:
Updated environment file generation command to match what laravel is calling the variable now

Commit:
Add support for PHPUnit 7.0

Commit:
Merge branch 'develop' into wip/laravel-5.9

Commit:
Merge branch 'develop' into wip/laravel-5.9

Commit:
Updated Composer dependencies for L6.

Temporarily removed the October subsplits until they are supporting L6.

Commit:
Fix test case incompatibilities, clean up ignores

Commit:
Fix incompatible `where` method in CmsObjectCollection

Signature for the `where` method changes in L6, so a wrapper has been put in place.

Commit:
Update unit tests to not rely on deprecated methods of testing

Commit:
Fix more exception tests

Commit:
Change module seeders to only temporary unguard attributes

Commit:
Seed admin password that passes our current validation rules

Commit:
Fix guarding for fixtures now that models aren't unguarded by seeders

Commit:
Un-mark skipped test - seems to be working

Commit:
Remove deprecated `getNotes` method from UpdateManager

Commit:
Ensure October helpers are loaded before Laravel helpers

Commit:
Fix missing output when UpdateManager is run within the Backend

Commit:
Link Input alias to October's Input facade

Commit:
Make slug route param optional for CMS module.

This allows the home page to load.

Commit:
Use updated branch name for library

Commit:
Drop PHP 7.1 tests

Commit:
Drop code reset and library replacement in automated tests, for now

Commit:
Fix code quality errors

LukeTowers commented on composer.json:
```diff
@@ -31,21 +31,19 @@
-        "october/backend": "~1.0",
-        "october/cms": "~1.0",
-        "laravel/framework": "^5.5.40",
+        "october/rain": "dev-wip/laravel-6",
```
```suggestion
        "october/rain": "dev-wip/laravel-6 as 1.0",
```
The backend modules require 1.0 so I think this is required @bennothommo (correct me if I'm wrong)

bennothommo replied:
@LukeTowers you're correct - however, I removed the modules from composer.json temporarily so I didn't have to deal with sub-splitting the branch. My plan was to re-add it back before merging it all in.

LukeTowers replied:
Ah, I see, makes sense

Commit:
Comment tweak for CmsObjectCollection::where()

Commit:
Move autoloading to index to make it the same as Laravel 6.

Implements work done in https://github.com/octobercms/october/pull/4280.

Credit to @Samuell1

Commit:
Partially revert and restore bootstrap/autoload.php

As discussed with @daftspunk, we now have to hard-code some overrides to ensure our helpers are loaded first. To keep the code DRY and self-explanatory, we do need to keep the autoload.php file.

LukeTowers commented on modules/cms/classes/CmsObjectCollection.php:
```diff
@@ -35,14 +36,38 @@ public function withComponent($components, $callback = null)
+     * @param  mixed  $value
+     * @return static
+     */
+    public function where($key, $operator = null, $value = null)
```
@bennothommo is this a breaking change technically? How does this affect the Halycon datasources?

bennothommo replied:
@LukeTowers I believe it's not a breaking change - the new method signature is technically less restrictive than the old signature. I've made that particular method a wrapper which replicates the restrictions of the old method and forwards the call on to a new method which is essentially our old method. `CmsObjectCollection` doesn't appear to be extended by another class, so I believe it should be fine.

LukeTowers replied:
@bennothommo what used to call this method though? Whatever called this method would still be sending the old signature. Additionally, it seems like this is misleading, since we haven't actually changed the behavior of the internal filter to take the operator into account.

bennothommo replied:
@LukeTowers I think it's mainly called when using a query against Halcyon objects. The above change will work with the old signature. It would only fail if someone wants to use the *new* signature for Halcyon objects (ie. adding an operator to `where` queries) - as far as I can tell, that's not happening as it's never been supported previously.

I felt that since Halcyon objects are something completely unique to October, it would be safe to keep the behaviour as is.

LukeTowers replied:
@bennothommo do we have to change the parameter names in order to pass the matching signature check? If not I'd prefer to keep the old names even if it means we have to change their defaults. 

Either way there should be a comment about how it doesn't actually match the new laravel signature

bennothommo replied:
@LukeTowers it appears we can keep the property names, but they have to remain with `null` as default in order to match the signature from Laravel. I've changed the method around to suit, and added a comment with the details of the change.

LukeTowers commented on modules/backend/models/User.php:
```diff
@@ -24,10 +24,10 @@ class User extends UserBase
+        'email' => 'required|between:6,191|email|unique:backend_users',
+        'login' => 'required|between:2,191|unique:backend_users',
+        'password' => 'required:create|between:8,191|confirmed',
+        'password_confirmation' => 'required_with:password|between:8,191'
```
@daftspunk @bennothommo I wonder if we need to go as deep as detecting what the default varchar length is with a special character to be parsed by the validation trait since we've introduced the config for it.

daftspunk replied:
Bit of a complex idea, 191 is a reasonable length. Note password does not need this limitation because the end result is hashed

LukeTowers replied:
Fair enough. @daftspunk what's the length of the hashed end result? Should we be concerned about 191 being too short? @bennothommo you can change this to be a min instead of a range.

daftspunk replied:
Password hash is 60 characters

bennothommo replied:
@LukeTowers I've changed the rules to use `min` instead. Should we do the same here: 
https://github.com/octobercms/library/blob/wip/laravel-6/src/Auth/Models/User.php#L27-L29
They are some pretty conservative rules.

LukeTowers replied:
@bennothommo hmm, I would say so yes. That might break some uses but nothing should really use the base model unmodified anyways

LukeTowers commented on modules/cms/routes.php:
```diff
@@ -22,7 +22,7 @@
      * handled by the back-end modules.
      */
-    Route::any('{slug}', 'Cms\Classes\CmsController@run')->where('slug', '(.*)?')->middleware('web');
+    Route::any('{slug?}', 'Cms\Classes\CmsController@run')->where('slug', '(.*)?')->middleware('web');
```
Why was this required?

bennothommo replied:
The home page wouldn't load without it, as the home page has no slug. :)

LukeTowers replied:
But home pages work fine right now, so what changed?

bennothommo replied:
This is what happens without the change:
![image](https://user-images.githubusercontent.com/15900351/72707513-8248cc00-3b9b-11ea-98c1-63b9f9f21fcf.png)
I can only surmise that Laravel has fixed something with regards to the difference between required URL parameters and optional ones.

LukeTowers replied:
Hmm, odd. Oh well, should be fine.

LukeTowers commented on tests/unit/system/classes/AutoDatasourceTest.php:
```diff
@@ -7,7 +7,7 @@
 class CmsThemeTemplateFixture extends Model
 {
-    protected $fillable = ['*'];
+    protected $guarded = [];
```
What was the reason for this change?

bennothommo replied:
`protected $fillable = ['*']` is actually incorrect. What happened was in our unit tests, when the seeders run on set up, they set all Eloquent models to be unguarded, so this was never picked up. However, Laravel must've improved their guarding, because the unit test that tests fillable attributes was failing. I've set the guarding for that fixture to use the correct method for making a single model unguarded.

LukeTowers replied:
Perfect, thanks for the explanation!

Commit:
Add Composer scripts

Commit:
Revert "Add Composer scripts"

This reverts commit a1cfc2aa58c2fa88c186a4306381fdcf1cd4f269.

Commit:
Add JUST the Composer scripts

Commit:
Use old signature for `where` method in CmsObjectCollection.

Refs: https://github.com/octobercms/october/pull/4893#discussion_r368408407

Commit:
Change User rules to use min for character lengths

Commit:
Adjust User rules to limit lengths of emails and logins

Commit:
Use default October log path for system logs

Commit:
Default to using Bootstrap 3 pagination.

L5.6 introduced Bootstrap 4 validation. While it is likely our pagination is all custom and is using the Storm UI, just to be safe, we'll make sure pagination is using Bootstrap 3 classes in all circumstances to retain current functionality.

This should have no effect on simple pagination, which is overriden with our custom template on the next line.

Commit:
Add Discord URL

Commit:
Drop monkey-patch on mock objects generator.

Reverts https://github.com/octobercms/october/commit/72f85ebc558a8e6fe514f6496f5393e535bac41b. No longer needed with PHPUnit 8.

LukeTowers commented on composer.json:
```diff
@@ -28,24 +28,23 @@
-        "meyfa/phpunit-assert-gd": "1.1.0",
+        "fzaninotto/faker": "^1.9",
+        "phpunit/phpunit": "^8.0|^9.0",
+        "phpunit/phpunit-selenium": "dev-master",
```
Is this temporary?

bennothommo replied:
@LukeTowers I believe so - they seem to be on the cusp of releasing v8.0 which supports PHPUnit 8 and 9. TBH though, I'm not sure if our Selenium tests still work.

LukeTowers replied:
Perhaps you can add the selenium tests back into the unit testing automation?

bennothommo replied:
@LukeTowers I can give it a try - I recall these sort of tests requiring us to fire up a headless version of Chrome to run the tests on, which was painful to install. Were these tests ever part of the automated testing?

LukeTowers replied:
Not sure, but if we can make it work with github actions that would be pretty sweet. I know there's a couple of plugins that make integrating with Selenium / Dusk in October a bit better, perhaps we should just try and get the current tests working standalone (and publish instructions for doing so) and then later on we can revisit automating them.

bennothommo replied:
No worries, I'll add it as a task to the list. I'll leave this thread open as well to remind me.

Teranode commented:
Getting an error when using Signed urls:
[log.txt](https://github.com/octobercms/october/files/4091806/log.txt)

Following: https://laravel.com/docs/6.x/urls#signed-urls

I've tried using:
```
Route::get('episode/download/{id}', ['as' => 'download', 'middleware' => 'signed', 'Batotwo\Show\Http\Controllers\Shows\ShowController@downloadFile']);
```
Using `->name('download')` will generate an error so i have to use `'as' => 'download'` ive also tried middleware `->middleware('Spatie\UrlSigner\Laravel\Middleware\ValidateSignature')` and `'middleware' => 'Spatie\UrlSigner\Laravel\Middleware\ValidateSignature'`

Also used 'Illuminate\Routing\Middleware\ValidateSignature'

but still ended up with that error.

LukeTowers commented:
@Teranode huh, didn't know that feature existed! Pretty nifty. @bennothommo once we've got it properly supported in October do you think we should add it to our docs?

bennothommo commented:
@LukeTowers sounds good to me. I'll add it to the list.

Commit:
Add simple unit tests for HasOneThrough and HasManyThrough relations

Commit:
Add support for Postmark mail transport configuration

Teranode commented:
Artisan commands giving off errors:
`php artisan cache:clear` doesn't work
`php artisan cache:forget` does work

but gives off that error. On further inspecting you're suppose to specify a `key` or a `store`, when using it. When I don't it defaults to removing all keys and all stores and then giving that error.
![](https://i.imgur.com/z1NOckR.png)

rajakhoury commented:
@Teranode both commands are working fine on my ends, been using this this with the file and redis stores. Isn't this related to folder permissions?

bennothommo commented:
@Teranode I would check the permissions of your `storage` folder to determine if that is causing `php artisan cache:clear` to fail. `cache:forget` is for removing a single item from the cache, so a key and a store is meant to be defined for that command.

I've just tested both on my own local copy and not had an issue.

Commit:
Update composer file to use new subsplit repos

Commit:
Merge branch 'develop' into wip/laravel-6

LukeTowers commented:
To anyone interested, we've made a lot of progress and the L6 upgrade is now ready for active testing. Please review https://github.com/octoberrain/meta/blob/master/l6-upgrade-notes.md for information on how you can test the changes in your own projects and any changes that may be required in your code bases.

Teranode commented:
@LukeTowers 
Upon changing the composer package via:
```
"october/rain": "dev-wip/laravel-6 as 1.0",
"october/system": "dev-wip/laravel-6",
"october/backend": "dev-wip/laravel-6",
"october/cms": "dev-wip/laravel-6",
```

I get this:
![](https://i.imgur.com/4Rz9JkN.png)

It happens for every backend url except for the ones appended with: `system/settings/update`

bennothommo commented:
@Teranode I've only used and tested the file cache at the moment.

FlusherDock1 commented on modules/system/ServiceProvider.php:
```diff
@@ -94,6 +94,7 @@ public function boot()
         }
 
+        Paginator::useBootstrapThree();
         Paginator::defaultSimpleView('system::pagination.simple-default');
```
Can we finally convert html markup of pagination from Bootstrap 3 to 4 ?

LukeTowers replied:
@FlusherDock1 you can feel free to override this in your own plugins but changing this away from Bootstrap 3 will break existing sites using the paginate()->render().

FlusherDock1 replied:
@LukeTowers I respect that, but using this kind of logic we can't update everything because there will be always one website that can be broken because of the new changes. 

P.S. Bootstrap 4 was released 6 years ago, I don't think someone really using Bootstrap 3 codebase right now.

LukeTowers replied:
@FlusherDock1 there's no-one stopping you from using whatever markup you want in your own projects. I can tell you right now that there are active sites with millions of views annually that utilize the default Bootstrap 3 markup. If you want to use Bootstrap 4, then just do the following in your own plugins:

```php
Paginator::defaultView('pagination::bootstrap-4');
Paginator::defaultSimpleView('pagination::simple-bootstrap-4');
```

bennothommo replied:
@FlusherDock1 The `useBootstrapThree` call is mainly for the Backend, which uses the Storm UI and thus uses Bootstrap 3 classes. Changing this would break the pagination on the backend. As @LukeTowers has indicated, you are more than free to use Bootstrap 4 for your own plugins.

FlusherDock1 replied:
>If you want to use Bootstrap 4, then just do the following in your own plugins:
Paginator::defaultView('pagination::bootstrap-4');
Paginator::defaultSimpleView('pagination::simple-bootstrap-4');

Thanks, i didn't know about that.

>The useBootstrapThree call is mainly for the Backend, which uses the Storm UI and thus uses Bootstrap 3 classes. Changing this would break the pagination on the backend. As @LukeTowers has indicated, you are more than free to use Bootstrap 4 for your own plugins.

About backend bootstrap 3 I know, I just suggested changes only for frontend. 

NVM, i think this question is closed due to storm ui limitations

Commit:
[Laravel 6] Add support for Laravel Dusk tests (#4919)

Commit:
Tweak code block in README

Commit:
Drop PHPUnit Selenium requirement

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

Commit:
Revert Laravel Dusk changes (#4919)

The Browser tests will be made into a RainLab plugin. (https://github.com/rainlab/dusk-plugin)

Commit:
Fix autoload map for tests

Commit:
Install RainLab Dusk plugin for browser tests

Commit:
Remove Dusk from dev dependencies

Commit:
Install Dusk plugin before getting Composer deps

Commit:
Enable Composer scripts for browser tests

Commit:
Install Chrome 78 driver

LukeTowers commented on .github/workflows/tests.yml:
```diff
@@ -83,6 +83,12 @@ jobs:
           path: ${{ steps.composer-cache.outputs.dir }}
           key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.json') }}
           restore-keys: ${{ runner.os }}-composer-
+      - name: Install Dusk plugin
```
Why don't we pull this in via composer?

bennothommo replied:
@LukeTowers we definitely can once it's been added to Packagist. I just wanted to get it working first.

martendej commented:
> Artisan commands giving off errors:
> `php artisan cache:clear` doesn't work
> `php artisan cache:forget` does work
> 
> but gives off that error. On further inspecting you're suppose to specify a `key` or a `store`, when using it. When I don't it defaults to removing all keys and all stores and then giving that error.
> ![](https://camo.githubusercontent.com/9122300df7193bf2dfdb564734677c7dafaf1817/68747470733a2f2f692e696d6775722e636f6d2f7a314e4f636b522e706e67)

I have the same issue, I have traced it down to the Illuminate\Cache\Console\ClearCommand expecting a boolean return value on the flush() method of the cache driver, which the October\Rain\Halcyon\MemoryRepository (default?) does not return.

bennothommo commented:
@martendej Thanks for the info. I will investigate it further.

multiwebinc commented:
https://github.com/octobercms/october/blob/wip/laravel-6/modules/system/lang/en/validation.php needs to be updated with the changes made in https://github.com/laravel/laravel/blob/master/resources/lang/en/validation.php. There are several entries missing (e.g. `gt`, `gte`, `lt`, `lte`).

bennothommo commented:
Thanks for the report @multiwebinc.

bennothommo commented:
@Teranode signed URLs should now work as of https://github.com/octobercms/library/pull/456/commits/c258e666b9c66174e773a57317764209665f2703.

Commit:
Add new supported filesystems to config

Commit:
Add new validation type lang strings.

Commit:
Use latest Chrome driver.

We'll have to assume that GitHub Actions' Ubuntu image is continually updated with the latest Chrome, although there's every likelihood that these tests will sometimes fail if the versions aren't in sync.

If it becomes too much of a problem, remove these tests.

bennothommo commented:
@multiwebinc Language files for validation have been updated: https://github.com/octobercms/october/commit/9911c6b5817088a99ac65d9f2ac76346402e1d47

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

multiwebinc commented:
Excellent work, @bennothommo

Commit:
Add `predis` client as default in config

Commit:
Use October provider for Redis service

daftspunk committed:
Remove local notes support

daftspunk committed:
Set default logging to single, remove dusk config

daftspunk committed:
Restore default password setting

daftspunk committed:
Remove selenium tests

daftspunk committed:
Restore autoloader and bootstrapper

There was no real need to change this from what it was

daftspunk committed:
Update composer to suit latest changes

daftspunk committed:
Switch to October Rain assetic

Commit:
Add roles and updated PHP version to module Composer requirements

daftspunk committed:
Minor

Teranode commented:
Almost forgot, Laravel Horizon will not function if the code below is not present in the app.php file.

```
    /*
    |--------------------------------------------------------------------------
    | Application Environment
    |--------------------------------------------------------------------------
    |
    | This value determines the "environment" your application is currently
    | running in. This may determine how you prefer to configure various
    | services your application utilizes. Set this in your ".env" file.
    |
    */
    'env' => env('APP_ENV'),
```

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Use correct folder for getting screenshot artifacts from browser tests

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Remove storage/framework/cache/data directory.

See https://github.com/octoberrain/meta/commit/18d7724287a668a6f9540bf95e858396a622223a for more information.

datune committed:
Replace Illuminate Dumper with Symfony Dumper (#4998)

See https://github.com/laravel/framework/pull/25087

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Fix for Laravel 6 upgrade

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

Commit:
Drop assertContains calls in OctoberEnv test

Commit:
Drop browser tests from test suite

As much as I'd like to keep them, they're just too flaky right now.

Commit:
Add info about RainLab Dusk for functional testing

Commit:
Fix PHPUnit location

Commit:
Minor path fix

Commit:
Merge branch 'develop' into wip/laravel-6

# Conflicts:
#	composer.json

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Switch to using league/csv >= 9.1

Commit:
Remove leftover code

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Make ImportModel compatible with League/CSV v9

Commit:
Further compat fixes

Commit:
Merge branch 'develop' into wip/laravel-6

# Conflicts:
#	modules/system/classes/UpdateManager.php
#	modules/system/classes/VersionManager.php

Commit:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

Commit:
Update providers and alias to use October Validator service

Commit:
Fix deprecated PHPUnit calls in ControllerTest

Commit:
Merge branch 'develop' into wip/laravel-6

Commit:
Merge branch 'develop' into wip/laravel-6

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

LukeTowers committed:
Fix Media Library test for L6

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

bennothommo committed:
Add app.loadDiscoveredPackages config item

Refs: https://github.com/octobercms/library/pull/492

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

bennothommo committed:
Remove Dusk "don't discover" flag

Will be fixed instead by https://github.com/octobercms/library/pull/492 when it is merged.

LukeTowers commented on config/app.php:
```diff
@@ -129,6 +129,26 @@
+    |
+    | Set this to `true` to enable automatic loading of these packages. This
+    | will result in packages being loaded, even if the plugin using them is
+    | disabled.
```
```suggestion
    | disabled. This is NOT RECOMMENDED.
```

LukeTowers committed:
Update config/app.php

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

bennothommo committed:
Namespace the authentication in tests under 'backend.auth'

Prevents conflicts with unit tests that might use another auth system, ie. Passport.

Hat tip to @LukeTowers for pointing out my shame.

bennothommo committed:
Add sensitive field input (#5201)

A field widget that allows for entering of sensitive information that can be revealed at the user's request - ie. API keys, secrets.

When a sensitive field that has been previously populated is loaded again, a placeholder is used instead of the real value, until the user opts to reveal the value. The real value is loaded via AJAX.

Credit to @tomaszstrojny for the original implementation.

Replaces #5062. Fixes #5061, #1850, perhaps #1061.

Co-authored-by: Tomasz Strojny <tomasz@init.biz>
Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Add fallback alias for Illuminate\Support\Facades\Input

This will redirect any instances of the now-removed Input facade from Laravel to the Rain library's Input facade.

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

bennothommo committed:
Add "develop.allowDeepSymlinks" configuration option.

Refs: https://github.com/octobercms/library/pull/491

bennothommo committed:
Provide an accessor for the form widget of the Settings controller (#5212)

Co-authored-by: Luke Towers <github@luketowers.ca>

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

LukeTowers committed:
Add L6 as a dep to the modules to prevent people with older Laravel versions from accidentally pulling in the L6 update

LukeTowers committed:
restore original password min length value

LukeTowers commented on modules/backend/routes.php:
```diff
@@ -25,7 +25,7 @@
             'prefix' => Config::get('cms.backendUri', 'backend')
         ], function () {
-            Route::any('{slug}', 'Backend\Classes\BackendController@run')->where('slug', '(.*)?');
+            Route::any('{slug?}', 'Backend\Classes\BackendController@run')->where('slug', '(.*)?');
```
Is this a change that needs to be documented? It affected RainLab.Translate so the possibility exists that other plugins could be affected.

bennothommo replied:
@LukeTowers done - https://github.com/octoberrain/meta/commit/81507a443a9027fcb9a03bb82b2b2b3f7c1652fb.

LukeTowers commented on modules/cms/twig/DebugExtension.php:
```diff
@@ -8,7 +8,7 @@
 use Cms\Classes\ComponentBase;
 use Illuminate\Pagination\Paginator;
 use Illuminate\Support\Collection;
-use Illuminate\Support\Debug\HtmlDumper;
```
Does this require a class alias setup in the library?

bennothommo replied:
@LukeTowers good idea - https://github.com/octobercms/october/commit/0dee0e54b100a22cdd36647161e951af04effcce

bennothommo committed:
Alias Illuminate\Support\Debug\HtmlDumper to Symfony's HtmlDumper

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/laravel-6

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

bennothommo committed:
Fix path validation for CmsObjects & Assets (#5229)

Instead of using a temporary file to allow checks with `realpath()` to succeed, use the new `resolve_path()` helper added in https://github.com/octobercms/library/commit/a9a629b0123c38608c48181704577f25798e77f7 to replace the `realpath()` usage and allow for path resolution of files that don't exist yet.

This will allow new asset files to still be saved correctly, whilst still preventing paths outside of the assets directory to be used.

Replaces https://github.com/octobercms/october/commit/b1954e92586f33a345068d56b957d7e366a5ba71.

LukeTowers committed:
Merge branch 'develop' into wip/laravel-6

LukeTowers committed:
L6 finalizing for merging to develop
